### PR TITLE
Parse subscription details into persisted metadata

### PR DESCRIPTION
**Pull request by JorgeMucientes**
Part of: #10124 

### Description

Parses `SubscriptionDetails` values into `metadata` field so we can then sync any updated values for subscription products with the API

### Testing instructions

1. Open an existing subscription product
2. Edit either the price or the expiration date
3. Save the product
4. Check that the values have been correctly updated, either by checking the product in web wp-admin or refreshing the product list and checking the updated values are still there. 

 🔂  Repeat the same steps for a product variation and verify the variation values are synced

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt [hunk 1]
```diff
@@ -272,7 +272,11 @@ class VariationDetailViewModel @Inject constructor(
                 periodInterval = periodInterval ?: subscription.periodInterval,
                 length = updatedLength
             )
-            viewState = viewState.copy(variation = variation.copy(subscriptionDetails = updatedSubscription))
+            val updatedVariation = variation.copy(subscriptionDetails = updatedSubscription)
+            viewState = viewState.copy(
+                variation = updatedVariation,
+                isDoneButtonVisible = updatedVariation != variation
+            )
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt:279]: This is unrelated to the main changes of the PR. Fixes a small issue were we'd change the product variation values but we won't show the "Save" button upon editing any of the subscription details.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt [hunk 2]
```diff
@@ -445,8 +449,7 @@ class VariationDetailViewModel @Inject constructor(
     }
 
     fun onSubscriptionExpirationChanged(selectedExpirationValue: Int) {
-        val newLength = if (selectedExpirationValue == 0) -1 else selectedExpirationValue
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt:448]: This is a leftover and not needed anymore. Default value for subscription `length` is now null

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt [hunk 3]
```diff
@@ -407,6 +409,31 @@ fun Product.toDataModel(storedProductModel: WCProductModel? = null): WCProductMo
         return jsonArray.toString()
     }
 
+    // Add any other editable product metadata here.
+    // Currently, only subscription details metadata is editable from the app. The rest is read only
+    fun SubscriptionDetails.toMetadataJson(): String {
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt]: WDYT about moving the function to `SubscriptionDetails` itself to avoid duplication?

We can then have something like this:

```kotlin
Product(...) {
     // Add any other editable product metadata here.
    // Currently, only subscription details metadata is editable from the app. The rest is read only
    private fun metadataJson() = subscription.toMetadataJson()
    ....
    
    it.metadata = metadataJson() ?: ""
}
```
And the same for variations.

**Review by hichamboushaba — APPROVED** [review]
Great work @JorgeMucientes, it works well, I left one suggestion, but it's not a blocking point.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt (hunk 3 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt]: I had second thoughts about that first. Since `meta_data` fields are not only about subscription details. So "encapsulating/hiding" it as an extension function inside `SubscriptionDetails.kt` would make less obvious in the future, where metadata parsing is happening and where to add additional parsing. However, I think is a minor thing with the benefit of removing duplication. So, I'm up for the refactor 👍🏼 !

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt [hunk 4]
```diff
@@ -15,3 +19,28 @@ data class SubscriptionDetails(
     val trialLength: Int?,
     val oneTimeShipping: Boolean
 ) : Parcelable
+
+// Currently, only subscription details metadata is editable from the app. The rest is read only and thus
+// not included in the persisted metadata.
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt]: np, @JorgeMucientes this comment is out of place now, can we move it to a function in `Product` itself?

To make this easier, we can probably make the function `toMetadataJson` return the raw `JsonArray` itself, and handle the conversion to `String` in the function in `Product`/`Variation`, this way, if we need to add more metadata in the future, we'll be able to benefit from `JsonArray` helper functions `add` and `addAll`, this is a minor point, just a suggestion 🙂

Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt]: Good point about returning the JsonArray() for easier chaining in the future. Thanks for all the tips @hichamboushaba 👍🏼

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt]: Thanks for the changes @JorgeMucientes, :shipit:
